### PR TITLE
LiteSPI generator supporting AXI-Lite and wishbone, modes 1x and 4x

### PR DESCRIPTION
**Pull request by suarezvictor**
Uses the full database of supported SPI devices and checks some parameters from the .yml against the device capabilities, to raise exceptions if support is missing (i.e. 8X mode)

### CODE REVIEW

File: litespi/gen.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+
+#
+# This file is part of LiteSPI.
+#
+# Copyright (c) 2015-2022 Florent Kermarrec <florent@enjoy-digital.fr>
+# Copyright (c) 2022 Victor Suarez Rovere <suarezvictor@gmail.com>
+# Copyright (c) 2020 Xiretza <xiretza@xiretza.xyz>
```
Comment (enjoy-digital) [inline litespi/gen.py:8]: Xiretza and Stefan contributed features for LiteEth that are not present here, please just keep you and me here.

File: litespi/gen.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+from litex.soc.integration.soc_core import *
+from litex.soc.integration.builder import *
+from litex.soc.integration.soc import SoCRegion
+
+import inspect
+import litespi.modules
+from litespi.spi_nor_flash_module import SpiNorFlashModule
+
+# module database
+modules_dict = {name: cls for name, cls in inspect.getmembers(litespi.modules) if inspect.isclass(cls) and issubclass(cls, SpiNorFlashModule)}
```
Comment (enjoy-digital) [inline litespi/gen.py:47]: Can you improve presentation? (not more than 100 characters per line if possible).

File: litespi/gen.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+        # SoCMini ----------------------------------------------------------------------------------
+        SoCMini.__init__(self, platform, clk_freq=int(1e6), **soc_args)
+
+        # CRG --------------------------------------------------------------------------------------
+        self.submodules.crg = CRG(platform.request("sys_clock"), platform.request("sys_reset"))
+
+        device = core_config["device"]
+        mode = core_config["mode"]
+        assert mode[-1] == "x" #expected parameter should end in "x" (typically 1x or 4x)
```
Comment (enjoy-digital) [inline litespi/gen.py:97]: Please prefer something like this:` assert mode in ["x1", "x4]`, this is more direct, easier to read and avoid comments.

File: litespi/gen.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        if not device in modules_dict:
+            raise ValueError("Unsupported SPI device")
```
Comment (enjoy-digital) [inline litespi/gen.py:99]: Please print the device name when not supported.

File: litespi/gen.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        
+        requested_bus_width = int(mode[:-1])
```
Comment (enjoy-digital) [inline litespi/gen.py:101]: Please use a dict instead: 
requested_bus_width = {
    "x1" : 1,
    "x4  : 4,
}[mode]

**Comment by suarezvictor** [comment]
will do

File: litespi/gen.py (hunk 1 above)
Comment (suarezvictor) [reply litespi/gen.py]: done, just take into account I based this source on the LiteEth one

**Comment by enjoy-digital** [comment]
@suarezvictor: OK thanks, let's wait to do some re-integration tests to merge it.

**Comment by enjoy-digital** [comment]
Thanks @suarezvictor, this is merged. I'll experiment more with it in the next days.

**Comment by suarezvictor** [comment]
Good, I'd recommend careful testing, including the combination of all parameters since they were contemplated and may work (simulation/hardware, x1/x4, SDR/DDR, axi-lite/wishbone)

**Pull request merged**